### PR TITLE
Kids

### PR DESCRIPTION
**Pull request by boryana-ilieva**
In this PR we are introducing children into the model. In every period, agents in the model receive a child with a certain exogenously given probability. The evolution of kids ages is recorded in the state space

### CODE REVIEW

**Comment by boryana-ilieva** [comment]
@peisenha please have a look, I will adjust and enhance the remaining tests after addressing your comments.

**Comment by boryana-ilieva** [comment]
@peisenha I cleaned up. I hope it is easier to review now. Looking forward to your comments!

File: soepy/exogenous_processes/children.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    child_age_update_rule[np.where(covariates[:, 1] != 0)] = (
+        states[np.where(covariates[:, 1] != 0)][:, 6] + 1
+    )
+    # Age does not exceed 11.
+    child_age_update_rule[child_age_update_rule > max(KIDS_AGES)] = 11
+
+    return child_age_update_rule
+
+
+def gen_prob_child_vector(
```
Comment (SeBecker) [inline soepy/exogenous_processes/children.py]: Codacy Issue found: [Dangerous default value PROB_CHILD_VALUES (builtins.list) as argument](https://app.codacy.com/manual/OpenSourceEconomics/soepy/pullRequest?prid=4986648)

File: soepy/test/test_unit.py [hunk 2]
```diff
@@ -79,110 +84,240 @@ def test_unit_states_hard_code():
...
+        )
+
+        np.testing.assert_equal(sum(df.dropna()["Age_Youngest_Child"] != -1), expected)
+
+
+def test_gen_probability_vector():
+    """This test ensures that the construction of the vector of probabilities
+    governing the child bearing process works for random number of periods and
+    random vector lengths for the vector of probabilities."""
+    for i in range(10):
```
Comment (SeBecker) [inline soepy/test/test_unit.py]: Codacy Issue found: [Unused variable 'i'](https://app.codacy.com/manual/OpenSourceEconomics/soepy/pullRequest?prid=4986648)

File: soepy/test/test_unit.py [hunk 3]
```diff
@@ -79,110 +84,240 @@ def test_unit_states_hard_code():
...
+            expected,
+        )
+
+
+def test_no_children_prob_0():
+    """This test ensures that child age equals -1 in the entire simulates sample,
+    equivalent to no kid is ever born, if the probability to get a child is zero
+    for all periods"""
+
+    for i in range(4):
```
Comment (SeBecker) [inline soepy/test/test_unit.py]: Codacy Issue found: [Unused variable 'i'](https://app.codacy.com/manual/OpenSourceEconomics/soepy/pullRequest?prid=4986648)

File: soepy/test/test_unit.py [hunk 4]
```diff
@@ -1,15 +1,20 @@
...
 
 import numpy as np
+from random import randrange, randint
 
 from soepy.pre_processing.model_processing import read_model_spec_init
 from soepy.pre_processing.model_processing import read_model_params_init
 from soepy.solve.solve_auxiliary import pyth_create_state_space
 from soepy.simulate.simulate_python import simulate
+from soepy.shared.shared_constants import LAST_CHILD_BEARING_PERIOD
+from soepy.exogenous_processes.children import gen_prob_child_vector
```
Comment (SeBecker) [inline soepy/test/test_unit.py]: Codacy Issue found: [Unused gen_prob_child_vector imported from soepy.exogenous_processes.children](https://app.codacy.com/manual/OpenSourceEconomics/soepy/pullRequest?prid=4986648)

File: soepy/test/random_init.py [hunk 5]
```diff
@@ -131,9 +131,17 @@ def random_init(constr=None):
     ) = np.random.uniform(0.1, 0.9, 3).tolist()
 
     (
-        model_params_init_dict["const_p"],
-        model_params_init_dict["const_f"],
-    ) = np.random.uniform(0.5, 5, 2).tolist()
+        model_params_init_dict["no_kids_f"],
```
[the remark this answers is not in the document]
Comment (boryana-ilieva) [reply soepy/test/random_init.py]: Can you give me some detail on that /mock line? I get the point, but I would not know how to set this up exactly.

**Comment by boryana-ilieva** [comment]
To complete before merge:

- [ ] #99 
- [ ] #95
- [ ] #94
- [ ] #93
- [ ] types in data frame
- [ ] update regression vault

**Comment by boryana-ilieva** [comment]
@peisenha can you help me with the following:

- more detail on how to improve `random_init` as requested in comment above
- general review of commits after last review
- any suggestion on how to resolve data frame types issue*

*Types are set correctly in line 179 of `simulate_auxiliary.py`. However all columns get converted back to float, when NaNs are added in the rows of individuals with higher levels of education who do not make choices in the first periods. I am thinking about the best way to resolve, but I don't have a suggestion I am happy with, yet.

File: .travis.yml [hunk 6]
```diff
@@ -20,32 +21,35 @@ before_script:
...
 
 notifications:
   slack:
-      rooms:
-        - oseconomics:Ft0TAdn9N8R38vMieCPxHV0a
-      on_success: never
-      on_failure: never
-
-  email: false
+    rooms:
```
Comment (peisenha) [inline .travis.yml]: We do not have slack set up anymore. Please delete whole block.

**Comment by peisenha** [comment]
@boryana-ilieva , here is what I had in mind for the missing data. Please see if this fits. https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html#integer-na

**Comment by peisenha** [comment]
Here is what I had in mind for the refactoring of the random specification:
soepy-refactoring

**Comment by boryana-ilieva** [comment]
> @boryana-ilieva , here is what I had in mind for the missing data. Please see if this fits. https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html#integer-na

Thanks, this is very helpful

**Comment by boryana-ilieva** [comment]
@peisenha, in the end two discussed changes remain uncompleted:
1. the refactoring of random init - suggested refactor did not result in the same type of list object, such that tests were failing. Since the issue concerns only code quality and time was very limited today, I'd prefer if we leave it for the detailed code review you will be doing in March.
2. dtypes - What you suggested works, in the sense that the targeted columns do get to be int. However the change breaks one of the unit tests that I generally consider a good test and I am out of time to look into this right now. I'd be thankful if we can record this as an issue and address on next PR.

I updated pandas and generated the regression vault anew. In case you consider any further changes urgently necessary, I will revert the last couple commits, that is no problem.

Talk to you tomorrow!

**Pull request merged**